File: android/Detonator/src/main/java/com/iconshot/detonator/element/AudioElement.java
package com.iconshot.detonator.element;

import android.net.Uri;
import android.os.Handler;
import android.view.View;

import androidx.media3.common.MediaItem;
import androidx.media3.exoplayer.ExoPlayer;

import com.iconshot.detonator.Detonator;
import com.iconshot.detonator.helpers.CompareHelper;

public class AudioElement extends Element<View, AudioElement.Attributes> {
    private ExoPlayer player;

    private final Handler handler = new Handler();
    private Runnable progressRunnable;

    private int currentPosition = 0;

    public AudioElement(Detonator detonator) {
        super(detonator);
    }

    @Override
    public Class<Attributes> getAttributesClass() {
        return Attributes.class;
    }

    @Override
    public View createView() {
        View view = new View(detonator.context);

        startTrackingProgress();

        setRequestListener("com.iconshot.detonator.ui.audio::play", (promise, value) -> {
            if (player == null) {
                promise.reject("No player available.");

                return;
            }

            player.play();

            promise.resolve();
        });

        setRequestListener("com.iconshot.detonator.ui.audio::pause", (promise, value) -> {
            if (player == null) {
                promise.reject("No player available.");

                return;
            }

            player.pause();

            promise.resolve();
        });

        setRequestListener("com.iconshot.detonator.ui.audio::seek", (promise, value) -> {
            Integer position = detonator.decode(value, Integer.class);

            if (player == null) {
                promise.reject("No player available.");

                return;
            }

            player.seekTo(position);

            promise.resolve();
        });

        return view;
    }

    @Override
    public void patchView() {
        String source = attributes.source;

        String prevSource = prevAttributes != null ? prevAttributes.source : null;

        boolean patchSource = forcePatch || !CompareHelper.compareObjects(source, prevSource);

        if (patchSource) {
            patchSource(source);
        }

        Boolean muted = attributes.muted;

        Boolean prevMuted = prevAttributes != null ? prevAttributes.muted : null;

        boolean patchMuted = forcePatch || !CompareHelper.compareObjects(muted, prevMuted);

        if (patchMuted || patchSource) {
            if (player != null) {
                if (muted != null && muted) {
                    player.setVolume(0);
                } else {
                    player.setVolume(1);
                }
            }
        }
    }

    @Override
    protected void patchDisplay(String display) {
        super.patchDisplay("none");
    }

    private void patchSource(String source) {
        if (player != null) {
            player.release();
        }

        if (source == null) {
            player = null;

            return;
        }

        player = new ExoPlayer.Builder(detonator.context).build();

        player.addListener(new ExoPlayer.Listener() {
            @Override
            public void onPlaybackStateChanged(int playbackState) {
                switch (playbackState) {
                    case ExoPlayer.STATE_ENDED: {
                        sendHandler("onEnd");

                        break;
                    }
                }
            }
        });

        Uri uri = Uri.parse(source);

        MediaItem mediaItem = MediaItem.fromUri(uri);

        player.setMediaItem(mediaItem);

        player.prepare();

        sendHandler("onReady");
    }

    @Override
    protected void removeView() {
        if (player != null) {
            player.release();
        }

        if (progressRunnable != null) {
            handler.removeCallbacks(progressRunnable);
        }
    }

    private void startTrackingProgress() {
        progressRunnable = new Runnable() {
            @Override
            public void run() {
                handler.postDelayed(this, 1000 / 30);

                int position = player != null ? (int) player.getCurrentPosition() : 0;

                if (position == currentPosition) {
                    return;
                }

                currentPosition = position;

                OnProgressData data = new OnProgressData();

                data.position = position;

                sendHandler("onProgress", data);
            }
        };

        handler.post(progressRunnable);
    }

    private static class OnProgressData {
        int position;
    }

    protected static class Attributes extends Element.Attributes {
        String source;
        Boolean muted;
    }
}

File: ios/Detonator/Sources/Detonator/Element/AudioElement.swift
import UIKit
import AVKit
import AVFoundation

class AudioElement: Element {
    private var player: AVPlayer?
    
    private var timer: Timer?
    
    private var currentPosition: Int = 0
    
    override public func decodeAttributes() -> AudioAttributes? {
        return super.decodeAttributes()
    }
    
    override public func createView() -> UIView {
        let view = UIView()
        
        startTrackingProgress()
        
        return view
    }
    
    override public func patchView() {
        let attributes = attributes as! AudioAttributes
        let prevAttributes = prevAttributes as! AudioAttributes?
        
        let source = attributes.source
        let prevSource = prevAttributes?.source
        
        let patchSourceBool = forcePatch || source != prevSource
        
        if patchSourceBool {
            patchSource(source: source)
        }
        
        let muted = attributes.muted
        let prevMuted = prevAttributes?.muted
        
        let patchMutedBool = forcePatch || muted != prevMuted
        
        if patchMutedBool || patchSourceBool {
            if let player = player {
                if muted == true {
                    player.volume = 0
                } else {
                    player.volume = 1
                }
            }
        }
    }
    
    override func patchDisplay(display: String?) {
        super.patchDisplay(display: "none")
    }
    
    private func patchSource(source: String?) {
        deinitPlayer()
        
        guard let source = source else {
            return
        }
        
        guard let playerURL = URL(string: source) else {
            return
        }
        
        self.player = AVPlayer(url: playerURL)
        
        self.initPlayer()
    }
    
    private func initPlayer() -> Void {
        let audioSession = AVAudioSession.sharedInstance()
        
        try? audioSession.setCategory(.playback, mode: .moviePlayback, options: [.mixWithOthers])
        try? audioSession.setActive(true)
        
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(playerDidEnd),
            name: .AVPlayerItemDidPlayToEndTime,
            object: player?.currentItem
        )
        
        emitHandler(name: "onReady")
    }
    
    private func deinitPlayer() {
        NotificationCenter.default.removeObserver(
            self,
            name: .AVPlayerItemDidPlayToEndTime,
            object: player?.currentItem
        )
        
        player?.pause()
        
        player = nil
    }
    
    override public func removeView() {
        deinitPlayer()
        
        timer?.invalidate()
    }
    
    public func play() throws -> Void {
        guard let player = player else {
            let error = NSError(domain: "com.iconshot.detonator.audio", code: -1, userInfo: [NSLocalizedDescriptionKey: "No player available."])
            
            throw error
        }
        
        player.play()
    }
    
    public func pause() throws -> Void {
        guard let player = player else {
            let error = NSError(domain: "com.iconshot.detonator.audio", code: -1, userInfo: [NSLocalizedDescriptionKey: "No player available."])
            
            throw error
        }
        
        player.pause()
    }
    
    public func seek(position: Int) throws -> Void {
        guard let player = player else {
            let error = NSError(domain: "com.iconshot.detonator.audio", code: -1, userInfo: [NSLocalizedDescriptionKey: "No player available."])
            
            throw error
        }
        
        let seconds = Double(position) / 1000
    
        let time = CMTime(seconds: seconds, preferredTimescale: 600)
        
        player.seek(to: time)
    }
    
    private func startTrackingProgress() {
        timer = Timer.scheduledTimer(withTimeInterval: 1 / 30, repeats: true) { timer in
            let time = self.player?.currentTime() ?? CMTime.zero
            
            let position = Int(CMTimeGetSeconds(time) * 1000)
            
            if position == self.currentPosition {
                return
            }
            
            self.currentPosition = position
            
            let data = OnProgressData(position: position)
            
            self.emitHandler(name: "onProgress", data: data)
        }
    }
    
    @objc func playerDidEnd() {
        emitHandler(name: "onEnd")
    }
    
    struct OnProgressData: Encodable {
        let position: Int
    }
    
    class AudioAttributes: Attributes {
        var source: String?
        var muted: Bool?
        
        required init(from decoder: Decoder) throws {
            let container = try decoder.container(keyedBy: CodingKeys.self)
            
            source = try container.decodeIfPresent(String.self, forKey: .source)
            muted = try container.decodeIfPresent(Bool.self, forKey: .muted)
            
            try super.init(from: decoder)
        }
        
        private enum CodingKeys: String, CodingKey {
            case source
            case muted
        }
    }
}
